### PR TITLE
Adding the ability to search with relational fields in the table. 

### PR DESCRIPTION
**Pull request by fawzanm**
This PR addresses the fix for issue #31. I have considered the comments on issue #31 and updated the PR with

1. Test Cases
2. Documentation

With this PR, we now have the ability to search on the relational models in the table. The notation used to distinguish between a normal field to search and a relational field is the `.`. 

eg. The key `name` will be a normal search field while `user.name` will be a field where we look to find the name on the `user` relationship of the `repository` in the `TableView`.

### CODE REVIEW

File: tests/Feature/ReviewTableViewTest.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        factory(FoodTest::class, 10)->create();
+        $this->users = factory(UserTest::class, 10)
+            ->create()
+            ->each(function (UserTest $user) {
+                $user->reviews()->saveMany(factory(ReviewTest::class, 1)->make());
```
Comment (Gustavinho) [inline tests/Feature/ReviewTableViewTest.php:32]: I think you don't need to specify the `1` to create just one object, without that argument should work out as well

File: tests/Feature/ReviewTableViewTest.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
+<?php
```
Comment (Gustavinho) [inline tests/Feature/ReviewTableViewTest.php:1]: I would rename the test file to something more general like 'RelationalSearchTest'

File: tests/Feature/ReviewTableViewTest.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            ->each(function (UserTest $user) {
+                $user->reviews()->saveMany(factory(ReviewTest::class, 1)->make());
+            });
+
+        $this->reviews = ReviewTest::all();
+
+        $this->livewire = Livewire::test(MockReviewTableView::class);
+    }
+
+    public function testSeeAllHeaders()
```
Comment (Gustavinho) [inline tests/Feature/ReviewTableViewTest.php:40]: I'm not sure about these methods, they are replicated in the table view tests, in these test file we want to be focus on the relational search, the `testSeeAllDataFoundBySearchInput ` might be enough, in that case we could make it even stronger with more test cases, what do you think?

File: src/Data/TableViewSearchData.php [hunk 4]
```diff
@@ -19,12 +19,34 @@ class TableViewSearchData implements Searchable
...
+
+                foreach ($regularFields as $field) {
                     $query->orWhere($field, 'like', "%{$value}%");
                 }
+
+                foreach ($relationalFields as $relationalValue) {
+
+                    [$relationship, $field] = explode('.', $relationalValue);
+
+                    $query->orWhereHas($relationship, static function ($query) use ($value, $field) {
```
[the remark this answers is not in the document]
Comment (fawzanm) [reply src/Data/TableViewSearchData.php:42]: If the relationship `foo` doesn't exist,  it throws an exception. ```Call to undefined method App\StudentPreference::foo()```, I don't think we should gracefully handle this, If we let Laravel throw the exception then the user will be able to track it easily. Let me know if you have any other suggestions for this.

File: tests/Feature/ReviewTableViewTest.php (hunk 1 above)
Comment (fawzanm) [reply tests/Feature/ReviewTableViewTest.php:32]: I actually had 10 there and had some issues when trying to assert while we have pagination. I will remove it.

File: tests/Feature/ReviewTableViewTest.php (hunk 2 above)
Comment (fawzanm) [reply tests/Feature/ReviewTableViewTest.php:1]: +1 This is a good suggestion.

File: tests/Feature/ReviewTableViewTest.php (hunk 3 above)
Comment (fawzanm) [reply tests/Feature/ReviewTableViewTest.php:40]: I had my doubts as well, I will remove the duplicate tests.

File: src/Data/TableViewSearchData.php [hunk 5]
```diff
@@ -19,12 +19,34 @@ class TableViewSearchData implements Searchable
...
+
+            $relationalFields = array_filter($fields, static function ($item) {
+                return str_contains($item, '.');
+            });
+
+            $regularFields = array_diff($fields, $relationalFields);
+
+            $query->where(static function ($query) use ($value, $regularFields, $relationalFields) {
+
+                foreach ($regularFields as $field) {
```
[the remark this answers is not in the document]
Comment (fawzanm) [reply src/Data/TableViewSearchData.php:34]: The query itself cannot be split but the `foreach` can be extracted. I will submit another PR.

[the remark this answers is not in the document]
Comment (fawzanm) [reply src/Data/TableViewSearchData.php:34]: Would something like this be okay? 

```
 /**
     * @param $relationalFields
     * @param $query
     * @param String $value
     */
    private static function applyRelationalFields($relationalFields, $query, string $value): void
    {
        foreach ($relationalFields as $relationalValue) {

            [$relationship, $field] = explode('.', $relationalValue);

            $query->orWhereHas($relationship, static function ($query) use ($value, $field) {
                $query->where($field, 'like', "%{$value}%");
            });
        }
    }

    /**
     * @param array $regularFields
     * @param $query
     * @param String $value
     */
    private function applyRegularFields(array $regularFields, $query, string $value): void
    {
        foreach ($regularFields as $field) {
            $query->orWhere($field, 'like', "%{$value}%");
        }
    }
```

File: src/Data/TableViewSearchData.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gustavinho) [reply src/Data/TableViewSearchData.php]: Oh great, I agree with you then

File: src/Data/TableViewSearchData.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (Gustavinho) [reply src/Data/TableViewSearchData.php]: yes exactly, just in case we have to add more complicated logic

**Comment by fawzanm** [comment]
Fixes based on the comments provided by @Gustavinho

**Pull request closed without merging**